src/dht_proxy_server.cpp

```diff
@@ -1050,6 +1009,46 @@ DhtProxyServer::handleCancelPushListen(const asio::error_code &ec, const std::st
         pushListeners_.erase(pushListener);
 }
 
+bool
+DhtProxyServer::handlePushListen(const InfoHash infoHash, const std::string pushToken, PushType type, const std::string clientId,
+                                 const std::shared_ptr<DhtProxyServer::PushSessionContext> sessionCtx, const std::string topic,
+                                 const std::vector<std::shared_ptr<Value>>& values, bool expired)
+{
+    Json::Value json;
+    json["key"] = infoHash.toString();
+    json["to"] = clientId;
+    using namespace std::chrono;
+    json["t"] = Json::Value::Int64(duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count());
+
+    {
+        std::lock_guard<std::mutex> l(sessionCtx->lock);
+        json["s"] = sessionCtx->sessionId;
+    }
+
+    // Build a comma-separated list of ids from the values. This is intended to be used when
+    // streaming from the iOS notification extension in order to filter out unwanted values.
+    std::string ids;
+    for(const auto& value : values) {
+        if (!ids.empty()) ids += ",";
+        ids += std::to_string(value->id);
+    }
+    json["ids"] = ids;
+
+    // If message is expired copy the value ID to the `exp` field.
+    // This is presumably used by the Android notification system.
+    if (expired && values.size() < 2) {
+        json["exp"] = json["ids"];
+    }
```

Comment:
do we want to put the id in "ids" if values expired ?

Reply:
yes, but only for backwards compatibility for iOS